File: UI/src/app/services/expenseSevices/expense.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import {catchError, Observable, throwError} from 'rxjs';


@Injectable({
  providedIn: 'root'
})
export class ExpenseserviceService {

  private BASE_URL= "http://localhost:8081/app/expense";

  constructor(private http: HttpClient) { }

  postExpense(expenseDto:any):Observable<any>{
    return this.http.post( this.BASE_URL, expenseDto);
  }

  getAllExpenses():Observable<any>{
    return this.http.get( this.BASE_URL + "/all");
  }

  deleteExpense(id: any): Observable<any> {
    return this.http.delete(this.BASE_URL + `/${id}`).pipe(
    catchError(error=>{
      return throwError(error)
    })
    );
  }

  getExpenseById(id: number):Observable<any>{
    return this.http.get( this.BASE_URL + `/${id}`);
  }

  updateExpense(id: number, expenseDto: any):Observable<any>{
    return this.http.put( this.BASE_URL + `/${id}`, expenseDto);
  }

}

File: Backend/src/main/java/lk/wara/proj/incomeexpensesorganizer/service/expense/ExpenseService.java
package lk.wara.proj.incomeexpensesorganizer.service.expense;

import lk.wara.proj.incomeexpensesorganizer.dto.ExpenseDto;
import lk.wara.proj.incomeexpensesorganizer.entity.Expense;

import java.util.List;

public interface ExpenseService {

    Expense postExpense(ExpenseDto expenseDto);

    List<Expense> getAllExpenses();

    Expense getExpenseById(Long id);

    Expense updateExpense(long id, ExpenseDto expenseDto);

    void deleteExpense(Long id);
}
